resources/tests/init.c

```diff
@@ -0,0 +1,44 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Init wrapper for boot timing. Alpine-specific because it points at /sbin/openrc-init.
+
+#include <sys/types.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+// Base address values are defined in arch/src/lib.rs as arch::MMIO_MEM_START.
+// Values are computed in arch/src/<arch>/mod.rs from the architecture layouts.
+// Position on the bus is defined by MMIO_LEN increments, where MMIO_LEN is
+// defined as 0x1000 in vmm/src/device_manager/mmio.rs.
+#ifdef __x86_64__
+#define MAGIC_MMIO_SIGNAL_GUEST_BOOT_COMPLETE 0xd0000000
+#endif
+#ifdef __aarch64__
+#define MAGIC_MMIO_SIGNAL_GUEST_BOOT_COMPLETE 0x40000000
+#endif
+
+#define MAGIC_VALUE_SIGNAL_GUEST_BOOT_COMPLETE 123
+
+int main () {
+   int fd = open("/dev/mem", (O_RDWR | O_SYNC | O_CLOEXEC));
```

Comment:
If we are going to integrate this into the `initrd.img` file, we need to make sure that `/dev/mem` exists when it's executed.

If it doesn't exist, we need to create it: `mknod("/dev/mem", S_IFCHR, makedev(1, 1));`

I updated the `initrd.img` with a modified version of this executable that calls `mknod` first, but the boot time seems to be quite long - about 1.2s. I haven't finished investigating what's the cause for the long boot time, but if nothing comes up, I can continue on Tuesday.

Reply:
Ok,  so yeah,  this looks a whole load more complicated than I thought :-(

This [post](https://unix.stackexchange.com/questions/52713/how-to-populate-dev-directory-when-building-my-own-initrd) sort of suggests that an `/init` wrapper containing `mount -t devtmpfs none /dev` before calling `/sbin/init` (or `/sbin/boottime_init` in our case) might be an option.   Don't know that would be quicker than making a static device.

This might would work if we just had a single boot time mechanism,  but I'm struggling to see how it would work for trying to maintain two.  AFACIT,  the `init` parameter gets ignored when loading an `initrd` image,  and `/init` is called no matter what.  Unless I'm doing something wrong.

Reply:
Why not stick with the simple solution we had before?

Directly bake and use the `init` wrapper in the image. The wrapper writes to the magic addr then runs the real `init`.

Simplifying assumptions:
1. Since we need the wrapper tool baked in the image, just place it as the default `init`, avoid using the `init=` boot param. Even when boot time is not measured, it only cost a few extra cycles at userspace initialization.
2. Since we control the image, use one that has `/dev/mem` set up right at boot. Don't add the extra complexity of handling mounting `/dev/mem`. We do need to continue gracefully if `/dev/mem` is not present (just run _real_ `init`).

@gc-plp @tnotw what do you think?

Reply:
Simple sounds good to me :-)

Reply:
@acatangiu, in the current initrd the first assumption is already in place, so even if we don't specify a custom `init=` the magic number will still be written.

We can also use a custom kernel to simplify things. I tested the idea by adding:
```diff
diff --git a/init/main.c b/init/main.c
index 0ee9c6866ada..31a4207dd20c 100644
--- a/init/main.c
+++ b/init/main.c
@@ -1001,6 +1001,11 @@ static int __ref kernel_init(void *unused)
 
        rcu_end_inkernel_boot();
 
+       void *base;
+       base = ioremap(0xd0000000, 0x4);
+       iowrite8(123, base);
+       iounmap(base);
+
        if (ramdisk_execute_command) {
                ret = run_init_process(ramdisk_execute_command);
                if (!ret)
```

in the 4.14 kernel just before calling `/init` and it works, with the following results:
- `Boot time with no network is: 86636 us`
- `Boot time with network configured is: 86456 us`
- `Boot time with initrd is: 146258 us`

The advantage of this method is that we don't need to patch multiple images to measure boot time (e.g. `number of architectures`x`number of images) and it's distro agnostic.
What do you think?

Reply:
Yes, that works too. But it just moves the burden of managing rootfs images to managing kernel images.
It can also be considered _cheating_ since that's still _kernel boot code_ 😝 .

**Pro**: we generally have more rootfs images than kernel images so fewer objects to maintain
**Con**: it is much easier to update rootfs images (simple file operations) than to patch and build kernel images

I personally lean toward doing it in userspace.

Reply:
@tnotw After a bit of struggling with packing the initrd image, it seems to work now. Apparently I was gzipping the initrd image, which introduced extra boot time.

Could you revert the `boot_args` parameter addition? Sorry for the back and forth.

Reply:
Sure, no problem.  Every day is a school day, as they say :smile: